Assets/Scripts/PlayerScripts/PlayerAction.cs

```diff
@@ -4,93 +4,76 @@
-            }
-        }
-    }
+    public abstract void SetTimeStamps(IEnumerable<Note> array);
 
     // Update is called once per frame
-    private void Update()
-    {
-        if (Time.time > 5 && !GameManager.current.IsGamePaused() && _inputIndex < timeStamps.Count)
-        {
-            _timeStamp = timeStamps[_inputIndex];
-            _marginOfError = MusicPlayer.current.marginOfError;
-            _audioTime = MusicPlayer.GetAudioSourceTime() - (MusicPlayer.current.inputDelayInMilliseconds / 1000.0);
-
-            if (Input.GetKeyDown(input))
-            {
-                GetAccuracy();
-            }
-
-            if (_timeStamp + _marginOfError <= _audioTime)
-            {
-                Miss();
-                print($"Missed {_inputIndex} note - time: {_timeStamp} audio time {_audioTime}");
-                _inputIndex++;
-            }
-        }
-    }
-
+    public abstract void Update();
 
-    private void GetAccuracy()
+    protected int GetAccuracy(double timeStamp, int inputIndex)
     {
-        if (Math.Abs(_audioTime - (_timeStamp)) < _marginOfError)
+        if (Math.Abs(_audioTime - (timeStamp)) < _marginOfError)
         {
             Hit();
-            print($"Hit on {_inputIndex} note - time: {_timeStamp} audio time {_audioTime}");
-            _inputIndex++;
+            print($"Hit on {inputIndex} note - time: {timeStamp} audio time {_audioTime}");
+            inputIndex++;
         }
         else
         {
             Inaccurate();
             print(
-                $"Hit inaccurate on {_inputIndex} note with {Math.Abs(_audioTime - _timeStamp)} delay - time: {_timeStamp} audio time {_audioTime}");
+                $"Hit inaccurate on {inputIndex} note with {Math.Abs(_audioTime - timeStamp)} delay - time: {timeStamp} audio time {_audioTime}");
         }
+        return inputIndex;
     }
 
-    private void Hit()
+    protected List<double> AddNoteToTimeStamp(Note cur_note, Melanchall.DryWetMidi.MusicTheory.NoteName cur_noteRestriction, List<double> cur_timeStamps){
+        if (cur_note.Octave == 1 && cur_note.NoteName == cur_noteRestriction)
+        {
+            var metricTimeSpan =
+                TimeConverter.ConvertTo<MetricTimeSpan>(cur_note.Time, MusicPlayer.MidiFileTest.GetTempoMap());
+            var spawnTime = ((double)metricTimeSpan.Minutes * 60f + metricTimeSpan.Seconds +
+                                (double)metricTimeSpan.Milliseconds / 1000f);
+
+            cur_timeStamps.Add(spawnTime - prespawnWarningSeconds);
+        }
+        return cur_timeStamps;
+    }
+
+    protected int CheckMiss(int inputIndex, List<double> cur_timeStamps, double cur_timeStamp){
+
+            if (cur_timeStamp + _marginOfError <= _audioTime)
+            {
+                Miss();
+                print($"Missed {inputIndex} note - time: {cur_timeStamp} audio time {_audioTime}");
+                inputIndex++;
+            }
+        return inputIndex;
+    }
+
+    protected void Hit()
```

Comment:
Minor issue: I think Hit() Miss() and Inaccurate() can be private since they are only used in this class